src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,208 @@
+                }
+
+                string key = reader.GetString()!;
 
-            if (metadata == MetadataPropertyName.Id)
+                // todo: verify value is converter.TypeToConvert and throw JsonException? (currently no test)
+                state.Current.ReturnValue = state.ReferenceResolver.ResolveReferenceOnDeserialize(key);
+                state.Current.ObjectState = StackFrameObjectState.MetadataRefPropertyEndObject;
+            }
+            else if (state.Current.ObjectState == StackFrameObjectState.MetadataIdProperty)
             {
-                // Special case for dictionary properties since those do not push into the ReadStack.
-                // There is no need to check for enumerables since those will always be wrapped into JsonPreservableArrayReference<T> which turns enumerables into objects.
-                object value = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
-                    state.Current.JsonPropertyInfo!.GetValueAsObject(state.Current.ReturnValue)! :
-                    state.Current.ReturnValue!;
+                if (!reader.Read())
+                {
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.String)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                }
 
-                state.ReferenceResolver.AddReferenceOnDeserialize(key, value);
+                state.Current.MetadataId = reader.GetString();
+
+                // Clear the MetadataPropertyName since we are done processing Id.
+                state.Current.JsonPropertyName = default;
+
+                if (converter.ClassType == ClassType.Enumerable)
+                {
+                    // Need to Read $values property name.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataValuesPropertyName;
+                }
+                else
+                {
+                    // We are done reading metadata.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataPropertyValue;
+                    return true;
+                }
             }
-            else if (metadata == MetadataPropertyName.Ref)
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefPropertyEndObject)
             {
-                state.Current.ReferenceId = key;
+                if (!reader.Read())
+                {
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.EndObject)
+                {
+                    // We just read a property. The only valid next tokens are EndObject and PropertyName.
+                    Debug.Assert(reader.TokenType == JsonTokenType.PropertyName);
+
+                    ThrowHelper.ThrowJsonException_MetadataReferenceObjectCannotContainOtherProperties(reader.GetSpan(), ref state);
```

Comment:
Pass the reader into the helper and call `GetSpan()` there rather than at the call site.

Reply:
This is called from two locations one of which doesn't need the reader. I don't think it's beneficial to create another throwhelper for that case.